### ISSUE TITLE
Azure Cloud Shell keeps resulting in "Key not found" error

### ISSUE THREAD

**Issue by rajhimanshu** [microsoft/terminal#7056]
I keep seeing the following behavior:

```
Tenant 0: Default Directory (brillfoxyahoo.onmicrosoft.com)
Tenant 1: Microsoft (microsoft.onmicrosoft.com)
Please enter the desired tenant number.
Enter n to login with a new account
Enter r to remove the above saved connection settings.
> 1
Requesting a cloud shell instance...
Succeeded.
Key not found
[process exited with code 1]
```

This issue seems persistent.

**Comment by Javanite** [comment]
I have the same issue. 

`To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code XXXXXXXX to authenticate.
This code will expire in 15 minutes.
Authenticated.
Tenant 0: xxx (xxx.onmicrosoft.com)
Tenant 1: xxx (xxx.onmicrosoft.com)
Tenant 2: xxx (xxx.onmicrosoft.com)
Tenant 3: xxx (xxx.onmicrosoft.com)
Please enter the desired tenant number.
> 0
Do you want to save these connection settings for future logins? [y/n]
> n
Requesting a cloud shell instance...
Succeeded.
Key not found
[process exited with code 1]`

**Comment by DHowett** [comment]
@Javanite I'm going to follow up with you offline, as I can furnish a debug build of the AzCS bridge :smile:

**Comment by iAmSaugata** [comment]
I am also having same issue.

Tenant 0: name
Please enter the desired tenant number.
Enter n to login with a new account
Enter r to remove the above saved connection settings.
> help
Please enter a valid number to access the stored connection settings, n to log in with a new account, or r to remove the saved connection settings.
> 0
Requesting a cloud shell instance...
Succeeded.
Key not found
[process exited with code 1]

**Comment by Javanite** [comment]
Hey Dustin,

Just a heads up I will be out the whole month of August, so I might be hard to reach. You can try pinging me on teams and I will respond if I can.

Mike


Mike_Kassis
| where Title == 'Senior PM Manager'
  and Company == 'Microsoft'
//Query like a boss - learn<https://docs.microsoft.com/en-us/azure/kusto/query/> KQL.

[MSFT_logo_Gray DE sized SIG1.png]

**Comment by yiigy** [comment]
Likely due to the shell storage hasn't been setup yet.  Login to your Azure Portal, select the Azure Shell on top to first initialize and setup the shell.  Once the storage is completed, you should be able connect via Azure Cloud Shell.

**Comment by Javanite** [comment]
@yiigy thanks for the suggestion, unfortunately doesn't appear to work. I started up the cloud shell in the portal, even ran it in interactive mode, but I still get the same result as before when using the Windows Terminal.

Here's a screencap of me running a random command in the az portal cloud shell.

![image](https://user-images.githubusercontent.com/19479604/88984582-cfaca100-d282-11ea-9253-912d62b1d291.png)

**Comment by csijake** [comment]
Having the same issue. Terminal Version: 1.2.2381.0

![image](https://user-images.githubusercontent.com/65626696/94025813-8bb7b300-fd7e-11ea-94b3-411badfbd318.png)

If these values matter:

`pwsh --version`: PowerShell 7.0.2
`$PSVersionTable`:

|Key                      |Value                       |
|-------------------------|----------------------------|
|PSEdition                |Core                        |
|PSCompatibleVersions     |System.Version[]            |
|Platform                 |Win32NT                     |
|GitCommitId              |7.0.2                       |
|OS                       |Microsoft Windows 10.0.18363|
|PSVersion                |7.0.2                       |
|PSRemotingProtocolVersion|2.3                         |
|SerializationVersion     |1.1.0.1                     |
|WSManStackVersion        |3.0                         |

**Comment by markashleybell** [comment]
I'm not sure if this is useful, but I was also seeing this issue, until I deleted the saved connection settings (`r`) and re-authenticated. I then saw a much more useful message, which led me to the actual cause:

![image](https://user-images.githubusercontent.com/483470/95451499-c5ccab80-095f-11eb-8527-636cf3af3d7d.png)

**Comment by jamesdecourcy** [comment]
I had the same problem. Going to https://shell.azure.com AND THEN changing from Bash to PowerShell.  This seemed to set everything up so it now works when I log in via Windows Terminal.

**Comment by wbpluto** [comment]
> Likely due to the shell storage hasn't been setup yet. Login to your Azure Portal, select the Azure Shell on top to first initialize and setup the shell. Once the storage is completed, you should be able connect via Azure Cloud Shell.

Yes, I know this helps. But why we cannot do this in Windows Terminal or give us a more clear message.

**Comment by Don-Vito** [comment]
@wbpluto - I have implemented a fix but was able to test it only once. Since I fixed the issue in azure portal by selecting a shell, I cannot reproduce the issue anymore. Even if I remove the user settings with an API or for new subscriptions.

My guess is that the problem happens only for "old" cloud shells, i.e., for those configured by an old versions of cloud console API (with API version lower than 2018-10-1). In such case, settings lack a "preferredShellType" field, that Terminal expects. As a result for such shells, Terminal succeeds to fetch the configuration, but then crashes trying to read this missing field.

I have added a code that prevents a crash and instead suggests to login in the browser and to set the shell.
I will try to push the code and let's see if the team likes it :blush:

**Comment by wbpluto** [comment]
> @wbpluto - I have implemented a fix but was able to test it only once. Since I fixed the issue in azure portal by selecting a shell, I cannot reproduce the issue anymore. Even if I remove the user settings with an API or for new subscriptions.
> 
> My guess is that the problem happens only for "old" cloud shells, i.e., for those configured by an old versions of cloud console API (with API version lower than 2018-10-1). In such case, settings lack a "preferredShellType" field, that Terminal expects. As a result for such shells, Terminal succeeds to fetch the configuration, but then crashes trying to read this missing field.
> 
> I have added a code that prevents a crash and instead suggests to login in the browser and to set the shell.
> I will try to push the code and let's see if the team likes it 😊

Thanks, so clear! Can you tell me how can I delete of get the Cloud Shell user setting with API or CLI?

**Comment by Don-Vito** [comment]
@wbpluto - after some work with the team, what we pushed was to fallback to powershell if "preferredShell" cannot be loaded.

The settings can be retrieved this way:
```
$token= ((Invoke-WebRequest -Uri "$env:MSI_ENDPOINT`?resource=https://management.core.windows.net/" -Headers @{Metadata='true'}).content |  ConvertFrom-Json).access_token
((Invoke-WebRequest -Uri https://management.azure.com/providers/Microsoft.Portal/usersettings/cloudconsole?api-version=2018-10-01 -Headers @{Authorization = "Bearer $token"}).Content | ConvertFrom-Json).properties | Format-List
```

and deleted this way:
```
$token= ((Invoke-WebRequest -Uri "$env:MSI_ENDPOINT`?resource=https://management.core.windows.net/" -Headers @{Metadata='true'}).content |  ConvertFrom-Json).access_token
Invoke-WebRequest -Method Delete -Uri https://management.azure.com/providers/Microsoft.Portal/usersettings/cloudconsole?api-version=2018-10-01 -Headers @{Authorization = "Bearer $token"}
```

**Comment by ChadLevy** [comment]
> I had the same problem. Going to [shell.azure.com](https://shell.azure.com) AND THEN changing from Bash to PowerShell. This seemed to set everything up so it now works when I log in via Windows Terminal.

![image](https://user-images.githubusercontent.com/909190/101399147-ad600b00-3894-11eb-9dec-ea15c6894069.png)

I had the same problem as well (obscure error message) and went to the same URL and was alerted that I needed to setup a storage account. After doing so (it was automatically handled by the site - just needed to specify which subscription) I was presented with the option to switch between Bash or PS.

Before selecting the option, I tried the cloud shell again and received a useful message:

> You have not set up your cloud shell account yet. Please go to https://shell.azure.com to set it up.

After selecting Powershell in the browser, I tried again in Terminal and it's working as expected.

**Comment by Don-Vito** [comment]
@ChadLevy - what version of the terminal are you using? We don't expect this error in the new builds.

**Comment by ChadLevy** [comment]
@Don-Vito 

1.4.3243.0

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito
> 
> 1.4.3243.0

@ChadLevy - Thanks! I was worried for a moment. The fix was introduced in 1.5.3142.0 (which is still in preview) and it will fallback to powershell without a need to go to shell.azure.com.

**Comment by homotechsual** [comment]
I reproduced this error on 1.16.2642.0 has the fix been reverted?

**Comment by zadjii-msft** [comment]
> has the fix been reverted

I don't believe there have been any meaningful changes to the azure cloud shell for some time now. You may want to file a new issue to track whatever you're seeing, as this thread is a couple years out of date now.

**Comment by coujean99** [comment]
Same problem here...
![image](https://user-images.githubusercontent.com/54377330/205214653-23c0e04d-cfd8-415e-8709-018acc43ca01.png)

**Comment by kororo211** [comment]
same issue with version 1.16.2642.0, any idea?
